handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java

```diff
@@ -229,4 +249,119 @@ protected void configurePipeline(ChannelHandlerContext ctx, String protocol) thr
             group.shutdownGracefully(0, 0, TimeUnit.MICROSECONDS);
         }
     }
+
+    @Test
+    public void testReplaceHandler() throws Exception {
+
+        assumeTrue(OpenSsl.isAvailable());
+
+        final String sniHost = "sni.netty.io";
+        LocalAddress address = new LocalAddress("testReplaceHandler-" + Math.random());
+        EventLoopGroup group = new DefaultEventLoopGroup(1);
+        Channel sc = null;
+        Channel cc = null;
+
+        SelfSignedCertificate cert = new SelfSignedCertificate();
+
+        try {
+            final SslContext sslServerContext = SslContextBuilder
+                    .forServer(cert.key(), cert.cert())
+                    .sslProvider(SslProvider.OPENSSL)
+                    .build();
+
+            final Mapping<String, SslContext> mapping = new Mapping<String, SslContext>() {
+                @Override
+                public SslContext map(String input) {
+                    return sslServerContext;
+                }
+            };
+
+            final Promise<Void> releasePromise = group.next().newPromise();
+
+            final SniHandler handler = new SniHandler(mapping) {
+                @Override
+                protected void replaceHandler(ChannelHandlerContext ctx,
+                        String hostname, final SslContext sslContext)
+                        throws Exception {
+
+                    boolean success = false;
+                    try {
+                        // SniHandler's replaceHandler() allows us to implement custom behavior.
+                        // As an example, we want to release() the SslContext upon channelInactive()
+                        // or rather when the SslHandler closes it's SslEngine. If you take a close
+                        // look at SslHandler you'll see that it's doing it in the #handlerRemoved0()
+                        // method.
+
+                        SSLEngine sslEngine = sslContext.newEngine(ctx.alloc());
+                        try {
+                            SslHandler customSslHandler = new SslHandler(sslEngine) {
+                                @Override
+                                public void handlerRemoved0(ChannelHandlerContext ctx) throws Exception {
+                                    try {
+                                        super.handlerRemoved0(ctx);
+                                        ReferenceCountUtil.release(sslContext);
```

Comment:
It's kind of on purpose. Should `super.handlerRemoved0(...)` throw it'll leave the refCnt at one and the test will fail at line ~354. Moving it into the finally will make the test "never" fail.

The hypothesis of this test is a little bit to show that you can now impl. a custom Sni/SslHandler and strategically override `handlerRemoved0()` which we know releases the SSLEngine. That failing would be really bad.